### PR TITLE
[4.6][BLOCKER]CLOUDSTACK-8883: Resolved connect/reconnect issue.

### PR DESCRIPTION
**Pull request by borisroman**
Hi!

@wilderrodrigues by implementing Callable you switched a couple of methods and fields. I switched them some more!

The reason why the Agent wouldn't reconnect was due to two facts.

Problem 1: Selector was blocking.
In the while loop at [1] _selector.select(); was blocking when the connection was lost. This means at [2] _isStartup = false; was never excecuted. Therefore at [3] the call to isStartup() always returned true resulting in an infinite loop.

Resolution 1: Move the call to cleanUp() [4] before checking if isStartup() has turned to false. cleanUp() will close() the _selector resulting in _isStartup to be set to false.

Problem 2: Setting _isStartup & _isRunning to true when init() throwed an unchecked exception (ConnectException).
The exception was nicely caught, but only logged. No action was taken! Resulting in _isStartup & _isRunning being set to true. Resulting in the fact the Agent thought it was connected successfully, though it wasn't.

Resolution 2: Adding return to the catch statement [5]. This way _isStartup & _isRunning aren't set to true.

Steps to test:
1. Deploy ACS.
2. Try all combinations of stopping/starting managment server/agent.

[1]https://github.com/borisroman/cloudstack/blob/b34f86c8d55a1cfc057585eab4db0fa2d98a7b3e/utils/src/main/java/com/cloud/utils/nio/NioConnection.java#L128
[2]https://github.com/borisroman/cloudstack/blob/b34f86c8d55a1cfc057585eab4db0fa2d98a7b3e/utils/src/main/java/com/cloud/utils/nio/NioConnection.java#L176
[3]https://github.com/borisroman/cloudstack/blob/b34f86c8d55a1cfc057585eab4db0fa2d98a7b3e/agent/src/com/cloud/agent/Agent.java#L404
[4]https://github.com/borisroman/cloudstack/blob/b34f86c8d55a1cfc057585eab4db0fa2d98a7b3e/agent/src/com/cloud/agent/Agent.java#L399
[5]https://github.com/borisroman/cloudstack/blob/b34f86c8d55a1cfc057585eab4db0fa2d98a7b3e/utils/src/main/java/com/cloud/utils/nio/NioConnection.java#L91

### CODE REVIEW

File: utils/src/main/java/com/cloud/utils/nio/NioConnection.java [hunk 1]
```diff
@@ -88,6 +88,7 @@ public void start() throws NioConnectionException {
             init();
         } catch (final ConnectException e) {
             s_logger.warn("Unable to connect to remote: is there a server running on port " + _port);
+            return;
```
Comment (wido) [inline utils/src/main/java/com/cloud/utils/nio/NioConnection.java]: Shouldn't we return on all Exceptions or just on this one?

Comment (borisroman) [inline utils/src/main/java/com/cloud/utils/nio/NioConnection.java]: The others throw checked exceptions which are caught upstream.

**Comment by remibergsma** [comment]
Thanks @borisroman will try to test soon!

**Comment by karuturi** [comment]
I tested this using redhat 6.3 as KVM hypervisor 
host is in UP state after stop/start of management server
Force reconnect of the host also worked.

:+1:

**Comment by pvr9711** [comment]
Thanks Boris, the functionality is working now- agent log shows reconnecting to MS when a connection is lost.  Let's merge this!

**Comment by karuturi** [comment]
ok. merging now.

**Pull request merged**